### ISSUE TITLE
Error in Torchsharp model inference in version 0.100.0

### ISSUE THREAD

**Issue by lindadamama** [dotnet/TorchSharp#999]
The ScriptModule loaded with version 0.99.6 of Torchsharp can run normally, while loading the same model with version 0.100.0 of Torchsharp keeps reporting errors
![err](https://github.com/dotnet/TorchSharp/assets/48863486/f24508de-cf51-4bdf-ac50-cd93894796d3)
Program running environment:
net6.0
Windows 10 nvdia mx450
Windows 11 nvdia  RTX4080
![err2](https://github.com/dotnet/TorchSharp/assets/48863486/b97bbe21-6fae-4f02-b599-cd8699130dc1)

**Comment by NiklasGustafsson** [comment]
What do the contents of the NuGet folder look like on your machine? It should be under your HOME folder:

C:\Users\YOURNAME\.nuget\packages\libtorch-cuda-11.7-win-x64-part6\2.0.1.1\runtimes\win-x64\native

The `nvrtc-builtins64_117.dll` file should be there.

If not, the safest thing is probably to wipe out all the .nuget\packages\*torch* folders and re-download the packages.

**Comment by NiklasGustafsson** [comment]
Also, look in the bin folder after you have built the code:

`bin\Debug\net6.0\runtimes\win-x64\native'

You should find the DLL there.

**Comment by lindadamama** [comment]
I have confirmed that all those DLLs exist and the path is also fine
![err3](https://github.com/dotnet/TorchSharp/assets/48863486/75594ad2-9d06-478f-bc48-24902753a235)

**Comment by ErrorGz** [comment]
![微信图片_20230519130143](https://github.com/dotnet/TorchSharp/assets/32073343/5fd34517-1b47-4b5d-a163-55497fed1c45)
I also encountered this message.

**Comment by NiklasGustafsson** [comment]
No, you should never have to install CUDA separately, just have the right drivers installed (and driver compatibility didn't change from 0.99.6 to 0.100.0). The necessary CUDA libraries come with the libtorch-cuda-windows NuGet package.

I will see if I can reproduced this locally today. I ran all the examples and tests on all platforms, with and without CUDA. I saw the warning, but not the exception. In the meantime, continue to use 0.99.6. It would be useful to know if this also happens on Linux, if anyone has access to it, or can run with WSL 2.

**Comment by NiklasGustafsson** [comment]
I have so far not been able to reproduce it. Since this seems to be related to TorchScript, would either one of you be willing to share the binary script file? You don't have to share any details about the model within, just tell me the shape of the input tensor(s).

**Comment by ErrorGz** [comment]
https://pan.baidu.com/s/1XeW4UX1v48HM1ojMo-5J_g?pwd=6xpt 
Thank you for your reply. I have uploaded the model file to Baidu Cloud Drive. Please download it. The shape of its input tensor is 3*416*416.

**Comment by lindadamama** [comment]
I'm testing the semantic segmentation model
<img width="542" alt="q" src="https://github.com/dotnet/TorchSharp/assets/48863486/22c03603-2b2d-4838-8e35-095d6ae748db">

**Comment by lindadamama** [comment]
[model.zip](https://github.com/dotnet/TorchSharp/files/11518238/model.zip)
1*3*600*800

**Comment by NiklasGustafsson** [comment]
Hmmm...

I was hoping for a clean repro, since that would have allowed me to debug rather than guess at the problem. However, no such luck. I first tried @lindadamama's model with PyTorch 2.0, wondering if this was a problem with version compat, but that wasn't it. Then, I tried it with a C# Notebook and TorchSharp, and I get the same output as PyTorch:

![image](https://github.com/dotnet/TorchSharp/assets/1791495/e863e1da-760a-41d6-897c-05b6303fb4e3)

However, I also don't see the warning when I run from the Notebook. I'll try with a VS project, instead.

**Comment by lindadamama** [comment]
I'm using     .net6.0 in vs2022

**Comment by NiklasGustafsson** [comment]
> I'm using .net6.0 in vs2022

Me, too:

```C#
{
    // Testing a bug.
   var input = torch.randn(8, 3, 600, 800);
   var m = torch.jit.load<Tensor, Tensor>(@"d:\Downloads\model\model.pt");
   m = m.to(torch.CUDA);
   input = input.to(torch.CUDA);
   var t = m.call(input);
}
```

From the debugger's immediate window:

```
t.ToString()
"[8x5x600x800], type = Float32, device = cuda:0"
```

Which is the same shape I get with PyTorch and with the C# Notebook. When I run from VS, I get the warning but not the exception.

Is there more than just calling the model that needs to be done before having it fail?

**Comment by lindadamama** [comment]
The model loading method seems to be different
<img width="266" alt="6" src="https://github.com/dotnet/TorchSharp/assets/48863486/74f74fb0-2c3a-4899-9fba-aa56c8bff9a1">
I don't know if it makes a difference

**Comment by NiklasGustafsson** [comment]
The typed load() only casts the output result statically, which makes no difference. I tried without it, and it works the same on my machine. I also added the 'eval()' call, but that made no difference. It still runs and produces a tensor of the right shape.

**Comment by NiklasGustafsson** [comment]
Does it work on CPU for you?

**Comment by NiklasGustafsson** [comment]
It's at the first call to `module.call()` (or `forward()`) that it fails, right?

**Comment by lindadamama** [comment]
yes

**Comment by NiklasGustafsson** [comment]
Can you try loading the model with PyTorch 2.0.1 and see if it runs on CUDA for you (it does for me, but so does it with TorchSharp)?

```Python
import torch
m = torch.jit.load("d:\\Downloads\\model\\model.pt").to('cuda')
m = m.eval()
t = torch.randn((1,3,600,800), device='cuda')
m(t)
```

**Comment by lindadamama** [comment]
pytorch2.0 is not yet installed on my computer. The version I installed is torch 1.12.1+cu113 cp39 cp39 win_ Amd64

**Comment by NiklasGustafsson** [comment]
@lindadamama -- could you go to the built application's (not the .nuget package's) `win-x64\native folder` and, using 'CMD' rather than the Windows Explorer, do a 'dir *.dll' and post the results here?

**Comment by lindadamama** [comment]
Is that it?
<img width="653" alt="666" src="https://github.com/dotnet/TorchSharp/assets/48863486/b18c49d4-2294-4390-8356-60aefe97341d">

**Comment by NiklasGustafsson** [comment]
Yes, and those files are a) the complete set of files from the libtorch distribution and b) the right file sizes, so you have everything that should be necessary. I can't tell what's wrong with the setup -- the fact that both you and @ErrorGz are seeing the same problem means it's not just you, but I'm unable to debug it since I can't reproduce it. 

I assumed that there was something wrong with the distribution, or a 1.12 / 2.0.1 or CUDA version incompatibility issue, but that's not the case, because it works on my machine (PyTorch, C# Notebook, Development-mode (not NuGet-base) Unit Test, and under TorchSharpExamples (NuGet-based)) with 2.0.1 and CUDA 11.7.

Further, since you module works with 0.99.6, which also is based on CUDA 11.7, driver incompatibility shouldn't be the issue. What driver are you using?

Here's what 'System Information' says about my RTX 2080 Super:

Driver Version	30.0.14.9613

**Comment by lindadamama** [comment]
I don't have my office computer around, so I can only reply to you on Monday. It's 1:19 in the morning here. I'm going to bed

**Comment by NiklasGustafsson** [comment]
I understand. Thanks for your patience!

**Comment by lindadamama** [comment]
It was an error on the second inference

**Comment by NiklasGustafsson** [comment]
So, should I call it twice, then?

**Comment by lindadamama** [comment]
first inference
<img width="371" alt="1" src="https://github.com/dotnet/TorchSharp/assets/48863486/b0083b26-5031-444a-8c00-930f826a29b7">
secend inference
<img width="647" alt="2" src="https://github.com/dotnet/TorchSharp/assets/48863486/ea00fe95-3734-400c-9141-dc1d095e26f2">

**Comment by lindadamama** [comment]
<img width="575" alt="image" src="https://github.com/dotnet/TorchSharp/assets/48863486/4e11a965-5174-467f-bffa-79460f9d5570">
Is there a problem with this?

**Comment by NiklasGustafsson** [comment]
What are the input tensor shape when it succeeds and when it fails? 

Maybe there's something in the new (2.0.1) runtime that makes it more sensitive to that?

**Comment by lindadamama** [comment]
<img width="577" alt="S1" src="https://github.com/dotnet/TorchSharp/assets/48863486/c8ec0aa7-4975-4be4-a24e-32ac39b853cd">
<img width="652" alt="S2" src="https://github.com/dotnet/TorchSharp/assets/48863486/6e4c0906-b52e-4cff-bcea-20d89bed3a9c">

**Comment by NiklasGustafsson** [comment]
It works for me, but I get a 1x5x530x704, not C = 7 as yours is showing above (results, not tensor).

**Comment by lindadamama** [comment]
I am testing images, and since the images on my computer are different from those in my office computer, the model is also different. The model I am testing now cannot be uploaded if it exceeds 35MB. The image size I trained for your model is 600 * 800

**Comment by NiklasGustafsson** [comment]
I see. We'll resolve this Monday -- it's after 2am where you are, isn't it?

**Comment by lindadamama** [comment]
yes

**Comment by lindadamama** [comment]
Has this situation recurred on your computer?This happened to all three of my computers

**Comment by NiklasGustafsson** [comment]
I have been unable to reproduce this issue on any of my machines, even though I have tried; I've only heard about this kind of issue from the two of you on this thread. 

There should not be any significant performance difference between 0.96.6 and 0.100.0.

**Comment by NiklasGustafsson** [comment]
If there's anybody else out there with a TorchScript model that fails in the same way, please let me know. I am very eager to reproduce this problem so that we can debug it and address it for real. Right now, I'm reduced to guessing... :-(

**Comment by lindadamama** [comment]
<img width="191" alt="屏幕截图 2023-05-24 215302" src="https://github.com/dotnet/TorchSharp/assets/48863486/caa1adb0-37c5-4862-a23d-00773d663733">
The project files, images, and models have been sent to your email. Please check them carefully

**Comment by NiklasGustafsson** [comment]
It's in my public profile:

https://github.com/NiklasGustafsson

**Comment by NiklasGustafsson** [comment]
>  The project files, images, and models have been sent to your email. Please check them carefully

Okay, I'll wait for them. If they don't fit on GitHub, I suspect that they are too big for email, too.

**Comment by lindadamama** [comment]
The email seems to have also failed, as the attachment is 43MB. Tomorrow, I will send you my small model including pictures

**Comment by lindadamama** [comment]
[Uploading ConsoleApp1.zip…]()
I have retrained a model, and both the model and images are in the bin folder

**Comment by ErrorGz** [comment]
> The email seems to have also failed, as the attachment is 43MB. Tomorrow, I will send you my small model including pictures

我觉得你可以发云盘😊

**Comment by NiklasGustafsson** [comment]
> Uploading ConsoleApp1.zip… I have retrained a model, and both the model and images are in the bin folder

That 'Uploading...' link just takes me back to this issue.

**Comment by lindadamama** [comment]
[ConsoleApp1.zip](https://github.com/dotnet/TorchSharp/files/11556445/ConsoleApp1.zip)
Uploaded successfully

**Comment by lindadamama** [comment]
Has the problem reoccurred?

**Comment by NiklasGustafsson** [comment]
Just downloaded it.

**Comment by NiklasGustafsson** [comment]
Here's what comes on the console when I run your model:

2.0.1.1
[W C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\jit\codegen\cuda\interface.cpp:47] Warning: Loading nvfuser library failed with: error in LoadLibrary for nvfuser_codegen.dll. WinError 126: The specified module could not be found.
 (function LoadingNvfuserLibrary)
0
1
2
3
4
????????3.2153134999999997?

The last seems to be the result of the next-to-last line in the main program. I don't think I have the right font installed in the console:

```C#
Console.WriteLine($"平均每张图片用时{stop.Elapsed.TotalSeconds / imgsPathss.Count()}秒");
```

**Comment by lindadamama** [comment]
Okay, that means there's no problem with the program. We still can't find the problem, but all three of my computers have the same problem, and the operating system language is Chinese

**Comment by NiklasGustafsson** [comment]
@ChengYen-Tang -- does that mean it's crashing? I can't decipher that.

**Comment by ChengYen-Tang** [comment]
@NiklasGustafsson ,
It didn't crash
<img width="637" alt="image" src="https://github.com/dotnet/TorchSharp/assets/32535494/2eae83f6-19f3-41e2-8aa9-0dfa33a4347f">

**Comment by ChengYen-Tang** [comment]
I can reproduce the problem @lindadamama  is having
<img width="1153" alt="image" src="https://github.com/dotnet/TorchSharp/assets/32535494/12cfbd03-f4f6-4008-952f-cb421505c82f">

**Comment by lindadamama** [comment]
I wonder if it's a problem with the C++version?

**Comment by ChengYen-Tang** [comment]
@NiklasGustafsson @lindadamama,
I found the problem.
Overwrite the dll in ConsoleApp1\ConsoleApp1\Inno.AIFrameWorkTest\bin\Debug\net6.0-windows\runtimes\win-x64\native with all dlls in the lib folder in the compressed file
https://download.pytorch.org/libtorch/cu118/libtorch-win-shared-with-deps-2.0.1%2Bcu118.zip

<img width="1051" alt="image" src="https://github.com/dotnet/TorchSharp/assets/32535494/ead3dd82-1969-4237-9f92-45efefbc6e0f">

**Comment by GeorgeS2019** [comment]
@ChengYen-Tang 
Can u explain, why u thought of libtorch118 while most of the discussion was about libtorch117

**Comment by lindadamama** [comment]
Did you use Cuda11.8?

**Comment by NiklasGustafsson** [comment]
No, 0.100.0 was built from the CUDA 11.7 PyTorch distribution. These were the ZIP files downloaded for the build:

```
05/10/2023  09:44     1,874,922,598 libtorch-shared-with-deps-2.0.1%2Bcu117.zip
05/10/2023  10:07     2,298,898,459 libtorch-win-shared-with-deps-2.0.1%2Bcu117.zip
05/10/2023  10:13     2,968,319,573 libtorch-win-shared-with-deps-debug-2.0.1%2Bcu117.zip
```

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson 
More of my ignorant, if the model.pt was trained using the same computer as the console program loading the model, would that reduce the incompatibility issue?

**Comment by lindadamama** [comment]
nvdia '531.79' 
rtx 3060
rtx 4080
mx450

**Comment by lindadamama** [comment]
After replacing the files, there's no problem on my end either

**Comment by NiklasGustafsson** [comment]
Hmm, if it were a 11.7 / driver compatibility issue, why does it work with 0.99.6? It was also built from a 11.7 libtorch distro.

This is very, very confusing.

**Comment by NiklasGustafsson** [comment]
> @NiklasGustafsson More of my ignorant, if the **model.pt** was trained using the same computer as the console program loading the model, would that reduce the incompatibility issue?

No, that shouldn't have anything to do with it. At least, I can't see how.

**Comment by ChengYen-Tang** [comment]
> @ChengYen-Tang Can u explain, why u thought of libtorch118 while most of the discussion was about libtorch117

I just downloaded a version from the pytorch official website for testing.

Here are my conclusions
1. The content of TorchSharp-cuda-windows may be missing some dlls or the dlls are damaged.
    I'll test it when I get home from get off work, or can someone help me test with the content of this zip file
https://download.pytorch.org/libtorch/cu117/libtorch-win-shared-with-deps-2.0.1%2Bcu117.zip

2. Driver conflicts

**Comment by NiklasGustafsson** [comment]
> The content of TorchSharp-cuda-windows may be missing some dlls or the dlls are damaged.

I have verified that there are no missing dlls, and I've been testing @lindadamama's model with the same NuGet package that is on the gallery, not a private collection of binaries. I guess that it is possible that they have been damaged during download, but I've never heard of such a thing happening before, and it seems very unlikely, since it repeats on several of @lindadamama's machines.

**Comment by NiklasGustafsson** [comment]
Reopening this issue -- I realize that there's a workaround, but I would still like to keep us digging at the problem until we have some idea of what is going on, so that it can be addressed or at least documented.

**Comment by lindadamama** [comment]
Okay, I will try my best to provide the information you need

**Comment by ChengYen-Tang** [comment]
@NiklasGustafsson @lindadamama @GeorgeS2019 ,

> The content of TorchSharp-cuda-windows may be missing some dlls or the dlls are damaged.
I'll test it when I get home from get off work, or can someone help me test with the content of this zip file
https://download.pytorch.org/libtorch/cu117/libtorch-win-shared-with-deps-2.0.1%2Bcu117.zip

I tested LibTorch cuda11.7 and the exception still occurs.
So it should be a driver issue.

Driver version: 535.50

**Comment by lindadamama** [comment]
My graphics card driver version is 531.79

**Comment by ChengYen-Tang** [comment]
Why is no one reporting this issue in pytorch

**Comment by NiklasGustafsson** [comment]
> Why is no one reporting this issue in pytorch

Yes, this whole thing is very mysterious, and it doesn't make sense that it would be a driver issue, either, since 0.99.6 is based on CUDA 11.7 too. 

I'd like to get to the bottom of it, but one possibility is to move to the CUDA 11.8-based libtorch and then make another release.

**Comment by ChengYen-Tang** [comment]
I hashed the 1.13.1 and 2.0.1 dlls.
<img width="1920" alt="image" src="https://github.com/dotnet/TorchSharp/assets/32535494/14132034-d358-42a3-a76e-ad2dac0bb57c">

**Comment by lindadamama** [comment]
The same graphics card driver version 531.79, I replaced the original DLL with Cuda11.8 DLL, and my 4080 graphics card still had the same error. The computer with the 3060 graphics card can run normally
Error with 4080 graphics card：
![4080](https://github.com/dotnet/TorchSharp/assets/48863486/41999234-fe2e-4bad-b4d7-3f134ed071ef)

**Comment by lindadamama** [comment]
Do you need to lower the version of the graphics card driver?

**Comment by NiklasGustafsson** [comment]
> May I need to use the lower version of the graphics card driver?

I don't know, I'm just grasping at straws here. 

I know I had problems with the newer drivers before and downgraded, but I don't recall it having anything to do with PyTorch / libtorch / CUDA.

**Comment by NiklasGustafsson** [comment]
Yup, that makes sense -- CUDA would either work or not work at all. The only thing I can think of is the driver version, but that doesn't make sense to me. 

Does overwriting with libtorch CUDA-11.8 binaries work for you? (See: https://github.com/dotnet/TorchSharp/issues/999#issuecomment-1561554522)

**Comment by NiklasGustafsson** [comment]
@nekomatic -- so overwriting with 11.8 does not work, then?

Regarding the CUDA toolkit -- that's a very interesting question. I haven't thought about that before, but I have 10.2, 11.5 and 11.7 all installed on my box, which does not exhibit this problem (neither of my two machines do). .NET is supposed to pick up in the local bin folder first, then in other places. I don't have any of the CUDA toolkits on my PATH, anyway, but maybe you do?

**Comment by NiklasGustafsson** [comment]
Wait a second -- I didn't pay attention to your first post. That warning is innocuous in and of itself. It's only if you later run into CUDA-related errors that you actually have the problem that was reported here.

**Comment by nekomatic** [comment]
I have just started to use TotchSharp, I didn't get far enough to get into any actual trouble. I've found this post following the traits of this warining, such warnings are better not left ignored... I still have a chance to select another techstack (either tensorflow-based or another version on torchsharp) so assumed better check if this has a porential to be a blocker.

**Comment by NiklasGustafsson** [comment]
>> I still have a chance to select another techstack

That is obviously your choice, and you should do so. 

As far as TorchSharp is concerned, v0.99.6, which is based on libtorch v1.13, does not show this warning. As far as the warning, it appears on Windows only, and appears to be related to the new `torch.compile` capabilities in v2.0, which TorchSharp does not yet support. There isn't a ton of downside to going with v1.13 (i.e. 0.99.6), since there isn't much new in v2.0 besides the compile functionality and bug fixes. That lets you get started with TorchSharp while this issue is figured out completely.

**Comment by nekomatic** [comment]
If it's Windows only then its not a problem, The target environment is linux so I will simply switch to WSL for the dev.

**Comment by NiklasGustafsson** [comment]
The quickest way to verify would be to go through the TorchSharp / CUDA tutorial with VS Code: https://github.com/dotnet/TorchSharpExamples/blob/main/tutorials/CSharp/tutorial5.ipynb

**Comment by lindadamama** [comment]
It may be a path problem. Today, I tried to copy the two DLLs that reported errors to the Bin folder, and the program can run normally
![123](https://github.com/dotnet/TorchSharp/assets/48863486/8d65ebb6-0fd1-42f8-897e-a92f76ca9076)

**Comment by NiklasGustafsson** [comment]
Fascinating! 

When I do the same, the warning disappears for me, too. (I didn't have any runtime problems, just the warning.)

I wonder if it's because the native libtorch runtime is loading this DLL dynamically, and only looks in the .EXE folder and the PATH and ignores looking relative to `libtorch.dll`

**Comment by lindadamama** [comment]
the native libtorch runtime is loading this DLL dynamically, and only looks in the .EXE folder and the PATH and ignores looking relative to libtorch.dll，I quite agree with your idea

**Comment by NiklasGustafsson** [comment]
I'm trying to think how we can easiest automate this copying of the DLLs into the .EXE folder... I guess it could be done when we're discovering the right libtorch.dll to load in TorchSharp.

https://github.com/dotnet/TorchSharp/blob/2bba4660a13b521add0795c3d24ffec5de9e65fa/src/TorchSharp/Torch.cs#L78

**Comment by NiklasGustafsson** [comment]
>> That would also explain the 1.13 vs. 2.0.1

It also explains why it's only happening when TorchSharp is loaded from the NuGet package -- when I build from source, all the libtorch dependencies ends up in the bin folder with the TorchSharp .dll and example .exe files.

**Comment by NiklasGustafsson** [comment]
Okay, so the fix is to preload 'nvfuser_codegen' from .NET, which will know about the package subfolders. Once loaded from managed code before it's loaded form native code, the path is not a problem. I'll do another release (0.100.2) and once I have verified it.

**Comment by NiklasGustafsson** [comment]
I believe this is fixed in version 0.100.2 -- I no longer get the warning. It has been released to NuGet.

**Comment by NiklasGustafsson** [comment]
@nekomatic, @lindadamama, @ErrorGz, @ChengYen-Tang -- I'm going to close this for now, but please reopen if it still fails the same way with 0.100.2.
